src/arch/riscv/pagetable_walker.cc

```diff
@@ -242,91 +284,355 @@ Walker::startWalkWrapper()
+            entry.gpte = entry.pte;
+            entry.pte = gresult.pte;
+
+
+            if (debug_condition()) {
+                DPRINTF(PageTableWalker,
+                                "[TWO-STAGE] Translated %#x -> %#x\n",
+                                vaddr, entry.paddr << PageShift |
+                                (vaddr & mask(entry.logBytes)));
+            }
+
+            if (!functional && !special_access) {
+                Addr vpn = getVPNFromVAddr(entry.vaddr, satp.mode);
+                walker->tlb->insert(vpn, entry);
+
+
+                if (debug_condition()) {
+                    DPRINTF(PageTableWalker,
+                        "[TWO STAGE] TLB Insert vaddr: %#x, pte: %#x\n",
+                            entry.vaddr , entry.pte);
+                    DPRINTF(PageTableWalker,
+                        "============================================\n");
+                }
+            }
+            else {
+                if (debug_condition() && functional) {
+                DPRINTF(PageTableWalker,
+                    "[FUNCTIONAL ACCESS] Translated %#x -> %#x\n",
+                    vaddr, entry.paddr << PageShift |
+                    (vaddr & mask(entry.logBytes)));
+                }
+
+                if (debug_condition() && special_access) {
+                    DPRINTF(PageTableWalker,
+                        "[SPECIAL ACCESS] NO TLB INSERT\n");
+                }
+
+                if (debug_condition())
+                    DPRINTF(PageTableWalker,
+                        "============================================\n");
+
+            }
+        }
+
         state = Ready;
         nextState = Waiting;
+
+        if (debug_condition()) {
+            DPRINTF(PageTableWalker, " Translated %#x -> %#x\n",
+            vaddr, entry.paddr << PageShift |
+            (vaddr & mask(entry.logBytes)));
+        }
     }
+
     return fault;
 }
 
 Fault
 Walker::WalkerState::startFunctional(Addr &addr, unsigned &logBytes)
 {
-    Fault fault = NoFault;
-    assert(!started);
-    started = true;
-    setupWalk(addr);
-
-    do {
-        walker->port.sendFunctional(read);
-        // On a functional access (page table lookup), writes should
-        // not happen so this pointer is ignored after stepWalk
-        PacketPtr write = NULL;
-        fault = stepWalk(write);
-        assert(fault == NoFault || read == NULL);
-        state = nextState;
-        nextState = Ready;
-    } while (read);
-    logBytes = entry.logBytes;
-    addr = entry.paddr << PageShift;
-
-    return fault;
+    // Pass the addess to entry here
+    // initState cannot because there is no req object
+    entry.vaddr = addr;
+    // just call walk
+    // it takes care to do the right thing
+    // when functional is true
+    return walk();
 }
 
 Fault
 Walker::WalkerState::stepWalk(PacketPtr &write)
 {
-    assert(state != Ready && state != Waiting);
+    int &curlevel = curstage == GSTAGE ? glevel : level;
```

Comment:
Why is this a reference?

Reply:
To change the actual object members when we write : curlevel--;